library/core/src/any.rs

```diff
@@ -202,10 +203,69 @@ pub trait Any: 'static {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T: 'static + ?Sized> Any for T {
     fn type_id(&self) -> TypeId {
+        unsafe impl<T: ?Sized + Any> OverrideAnyTypeId for T {
+            default fn overridden_type_id() -> TypeId {
+                TypeId::of::<Self>()
+            }
+        }
+
+        <T as OverrideAnyTypeId>::overridden_type_id()
+    }
+}
+
+// Safety: overriding the dynamic `type_id` of a type must not be done
+// in way where arbitrary code may witness the static and dynamic type mismatch.
+unsafe trait OverrideAnyTypeId: Any {
+    fn overridden_type_id() -> TypeId;
+}
+
+#[allow(missing_debug_implementations)]
+#[doc(hidden)]
+#[unstable(feature = "drop_no_unwind_same_any_type_id", issue = "none")]
+#[repr(transparent)] // repr ≥ C to ensure same layout as `T`
+pub struct DropNoUnwindSameAnyTypeId<T>(ManuallyDrop<T>);
+
+#[unstable(feature = "drop_no_unwind_same_any_type_id", issue = "none")]
+impl<T> DropNoUnwindSameAnyTypeId<T> {
+    /// # Safety
+    ///
+    /// The value returned by `new()` must be immediately type erased to a
+    /// `Ptr<dyn Any>` so that no code sees the mismatch between static and
+    /// dynamic types.
+    #[doc(hidden)]
+    #[unstable(feature = "drop_no_unwind_same_any_type_id", issue = "none")]
+    pub unsafe fn new(value: T) -> Self {
+        Self(ManuallyDrop::new(value))
+    }
+}
+
+// SAFETY: from the safety preconditions of `DropNoUnwindSameAnyTypeId::new()`.
+unsafe impl<T: 'static> OverrideAnyTypeId for DropNoUnwindSameAnyTypeId<T> {
+    fn overridden_type_id() -> TypeId {
         TypeId::of::<T>()
     }
 }
 
+#[unstable(feature = "drop_no_unwind_same_any_type_id", issue = "none")]
+impl<T> Drop for DropNoUnwindSameAnyTypeId<T> {
+    fn drop(&mut self) {
+        struct AbortOnDrop {}
+
+        impl Drop for AbortOnDrop {
+            fn drop(&mut self) {
+                crate::intrinsics::abort();
```

Comment:
This should probably use something like the `panic_no_unwind` added in https://github.com/rust-lang/rust/pull/92828 to call the panic handler again with a helpful message.

The alternative to a unwind-to-abort shim of course would be just to leak the panic payload.

Another would be to set the panic count before calling `ManuallyDrop::drop` such that any panics in the payload drop immediately abort as panic-during-unwind rather than waiting until we hit this landing pad, though this necessarily makes more executions abort.

Reply:
Yeah, I went with a good old simple abort for starters, but there is indeed room for improvement here 💯

I don't have preferences for this use case, although aborting (rather than, _e.g._, leaking within a pretty crazy scenario) seems fine; in that case, I like the simplicity of `panic_no_unwind` in this drop glue, although to be fair the panic-count idea has its appeal. Maybe it's just me, however, but I do have a slightly uneasy feeling with panic-count: it's kind of a bit too entrenched with other logic / I'd be worried with a change in panic-count semantics indirectly affecting this snippet.

So, all in all, I guess I have kind of a soft preference for:

```suggestion
                crate::panicking::panic_no_unwind();
```

Thoughts?

Reply:
Definitely seems reasonable, though perhaps you'd want a message other than `panic_no_unwind`'s hardcoded message for ~`#[nounwind]` functions. See also https://github.com/rust-lang/project-error-handling/issues/34 which is relevant.

The other other option would be just to trigger a normal panic-in-unwind directly here, like

```suggestion
                // if this gets hit, we're already unwinding, so this will immediately abort
                panic!("dropping unwind payload panicked");
```

though that does also rely on panic-count/panic-in-unwind semantics, which [_might_ not be guaranteed?](https://github.com/rust-lang/rust/blob/1c7b36d4db582cb47513a6c7176baaec1c3346ab/library/std/src/panicking.rs#L707-L714)

This'd be easier if this type was in std, because a) the previous could be written to use `catch_unwind` and call `abort` afterwards if it didn't already abort for some reason, or it could even just be

```suggestion
                rtabort!("dropping unwind payload panicked");
```

`rtabort!` is [what is done by `lang_start`](https://github.com/rust-lang/rust/blob/2681f253bcdb31a274411d2be456e7c6a1c67d62/library/std/src/rt.rs#L129-L132) when an unwind manages to escape containment.

Putting this in std also means that it can get a "graceful" process abort rather than the intrinsic abort which is just `ud2`. Since the wrapper is only used in `std`, perhaps it makes sense to move there so it can utilize the std-aware abort routine? Though this would mean exposing `OverrideAnyTypeId` from core to std instead of `DropNoUnwindSameAnyTypeId` which is marginally safer to work with.

Reply:
> Though this would mean exposing

Exactly: when I started this PR I went with the wrapper type living in `std::panicking` since it made more sense, but that meant _exposing the `Any`-overrides-through-specialization_, which is something I felt _very uneasy_ about: that's a can of worms that even behind a scary-looking feature-gate I'd rather keep private. Moving the wrapper to `core` then turned out to be an elegant way to achieve that, but indeed, at the cost of losing access to some of the convenience logic available in `std`, but I think it's better overall 🙂

  - I went with your panic-in-panic approach (since it "fits" the `catch_unwind` theme, honestly), and as your investigations / inquiries showed, this panic-in-panic=>abort behavior is so pervasive and prevalent that it's expected to be relied on, and it won't be changed without an opt-in (such as the one in your proposal: a function annotation)